docs/RFCS/20180411_alter_column_type.md

```diff
@@ -0,0 +1,387 @@
+existing column data, such as `FLOAT -> INT USING ceil(column_name)`.
+The allowable expressions are the same as those used in
+[computed columns](https://www.cockroachlabs.com/docs/stable/computed-columns.html#creation),
+namely pure functions using the data available in a single row.
+
+## Two-phase change
+
+An `ALTER ... COLUMN TYPE` command is implemented in two phases:
+1. Add a new, hidden, computed column to the table that effects the
+   conversion.
+2. Atomically remove the computed nature from the new column, swap it
+   for the original, and drop the original column.
+
+Given that backfilling the computed column may take an arbitrary and
+unpredictable amount of time, we will document and support a "manual"
+version of the above, allowing a user to control exactly when a type
+change becomes visible to SQL clients.
+
+## Key constraints
+
+Columns that are the target of a primary or foreign key constraint will
+not be supported under this RFC. It could be possible to support foreign
+keys in the future by cascading the type change, but this would greatly
+increase the amount of coordination required to effect the total schema
+change.
+
+### Interleaved tables
+
+Changing the types of primary keys is not within the scope of this RFC,
+so this feature does not interact with interleaved tables. When changes
+to primary keys are supported, a change to the PK of a parent table
+would have to be cascaded to the corresponding PK prefix column of any
+interleaved descendant tables.
+
+## Indexed columns
+
+Columns that are part of an index are supported, however columns that
+are part of a primary key are explicitly out of scope for this change.
+
+## Views
+
+Columns that are depended upon by a `CREATE VIEW` are out of scope for
+this RFC. A future change to support views should be possible in the
+same manner as supporting columns that are the target of a foreign-key
+constraint.
+
+# Reference-level explanation
+
+## Process
+
+Conceptually, a column type change can be thought of as creating an
+anonymous, "shadow" column of the new type, which captures writes to the
+original, source, column while a backfill process is run. Once the
+backfill is complete, the shadow column is swapped for the original,
+which is then retired. The shadow column will be generally invisible to
+the user until it is swapped in. Shadow columns will be created in the
+same column family as the source column.
+
+This list describes the general type-change case, not all of these steps
+are necessarily executed for trivial or validated changes.
+
+* `ALTER ... COLUMN TYPE` is run, resulting in a new shadow column being
+  created. The shadow column will be created as a computed column, in
+  the source column's family, with the expression provided by the user
+  via `USING expression` or inferred for cast-type conversions.
+  * If there are indexes that refer to the column, additional
+    shadow-indexes are created.
```

Comment:
Note for the implementation: be mindful that cockroachdb currently supports (for better or for worse) multiple indexes over the same set of columns, but with different names. Perhaps we shouldn't do that, but we can. There must be as many shadow indexes created as there were indexes originally, and the atomic rename described below must apply separately to all of them.